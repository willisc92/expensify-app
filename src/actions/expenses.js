import uuid from 'uuid';
import database from '../firebase/firebase'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// With firebase:
// component calls action generator
// action generator returns function
// component dispatches function
// function runs (has ability to dispatch other actions)


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid; 
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = {description, note, amount, createdAt};
        
        return database.ref(`users/${uid}/Expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense 
            }));
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id 
});

export const startRemoveExpense = ({id} = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/Expenses/${id}`).remove().then((ref) => {
            dispatch(removeExpense({id}));
        });
    };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/Expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid; 
        return database.ref(`users/${uid}/Expenses`)
            .once('value')
            .then((snapshot) => {
                const expenses = [];

                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch(setExpenses(expenses));
            });   
    };
};