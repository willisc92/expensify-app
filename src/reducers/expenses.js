// Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense]; // array spread operator syntax- equivalent to state.concat(action.expense).
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates   // object spread operator - copies/overrides key-values from object(s) into the new object
                        // requires "babel-plugin-transform-object-rest-spread" plugin
                    }
                } else {
                    return expense;
                }
            });
        default:
            return state;
    }
};

