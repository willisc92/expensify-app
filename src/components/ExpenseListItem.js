import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({id, dispatch, description, amount, createdAt}) => (
    <div>
        <h3>{description}</h3>
        <p>Amount: ${amount/100} - Created At: {createdAt}</p>
        <button onClick = {() => {
           dispatch(removeExpense({id})); 
        }}>Remove</button>
    </div>
);

export default connect()(ExpenseListItem);