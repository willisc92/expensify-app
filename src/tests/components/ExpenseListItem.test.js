import React from 'react';
import expenses from '../fixtures/expenses';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';

test('Should render ExpenseListItem', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});