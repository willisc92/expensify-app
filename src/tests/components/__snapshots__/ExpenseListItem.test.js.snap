// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListItem 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    Amount: $
    1.95
     - Created At: 
    0
  </p>
</div>
`;
